test/mpi/threads/pt2pt/mt_perf_pt2pt.c

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+/* This program provides a simple test of send-receive performance between
+   two (or more) processes.  This sometimes called head-to-head or
+   ping-ping test, as both processes send at the same time.
+*/
+
+#include "mpi.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include "mpitest.h"
+#include "mpithreadtest.h"
+
+#define DEF_NUM_THREADS 2
+
+#define MESSAGE_SIZE 8
+#define NUM_MESSAGES 64000
+#define WINDOW_SIZE 64
+
+#define PAGE_SIZE 4096
+
+#define ERROR_MARGIN 0.05       /* FIXME: a better margin? */
+
+MPI_Comm *thread_comms;
+double *t_elapsed;
+
+int runtime_num_threads = -1;
+
+MTEST_THREAD_RETURN_TYPE send_thread(void *arg);
+MTEST_THREAD_RETURN_TYPE recv_thread(void *arg);
+
+MTEST_THREAD_RETURN_TYPE send_thread(void *arg)
+{
+    int tid;
+    MPI_Comm my_comm;
+    int win_i, win_post_i, win_posts;
+    void *send_buf;
+    int sync_buf;
+    MPI_Request requests[WINDOW_SIZE];
+    MPI_Status statuses[WINDOW_SIZE];
+    double t_start, t_end;
+
+    tid = (int) (long) arg;
+    my_comm = thread_comms[tid];
+
+    win_posts = NUM_MESSAGES / WINDOW_SIZE;
+    if (win_posts * WINDOW_SIZE != NUM_MESSAGES)
+        printf("Warning: The final reported numbers will be inaccurate.\n");
+
+    /* Allocate a page-aligned buffer */
+    posix_memalign(&send_buf, PAGE_SIZE, MESSAGE_SIZE * sizeof(char));
```

Comment:
Alignment is critical. At the least, we need the buffers of different threads to be on different cache lines to prevent any potential serialization (either in the NIC or in the CPU) occurring from parallel access (reads or writes) to the same cache line. Experiments on both InfiniBand and Omni-Path have shown hurtful performance when the buffer is shared between the threads even for send operations. I am doing page alignment instead of cache alignment because I am paranoid.

Reply:
I think using `posix_memalign` is more helpful for the user since it is direct and explicit. Adding extra space in the buffer is also a solution, but it seems to be indirect and implicit.

Right, I can change the alignment to the size of a cache line instead of the page size.

Reply:
Added a comment in the latest version.